python/taichi/lang/linalg.py

```diff
@@ -81,6 +81,46 @@ def svd3d(A, dt, iters=None):
     return U, sigma, V
 
 
+@ti.func
+def eig2x2(A, dt):
+    tr = A.trace()
+    det = A.determinant()
+    gap = tr**2 - 4 * det
+    lambda1 = ti.Vector.zero(dt, 2)
+    lambda2 = ti.Vector.zero(dt, 2)
+    v1 = ti.Vector.zero(dt, 4)
+    v2 = ti.Vector.zero(dt, 4)
+    if gap > 0:
+        lambda1 = ti.Vector([tr + ti.sqrt(gap), ti.cast(0.0, dt)]) * 0.5
+        lambda2 = ti.Vector([tr - ti.sqrt(gap), ti.cast(0.0, dt)]) * 0.5
+        A1 = A - lambda1[0] * ti.Matrix.identity(dt, 2)
+        A2 = A - lambda2[0] * ti.Matrix.identity(dt, 2)
+        if all(A1 == ti.Matrix.zero(dt, 2, 2)) and all(
+                A1 == ti.Matrix.zero(dt, 2, 2)):
+            v1 = ti.Vector([0.0, 0.0, 1.0, 0.0]).cast(dt)
+            v2 = ti.Vector([1.0, 0.0, 0.0, 0.0]).cast(dt)
+        else:
+            v1 = ti.Vector([A2[0, 0], 0.0, A2[1, 0],
+                            0.0]).cast(dt).normalized()
+            v2 = ti.Vector([A1[0, 0], 0.0, A1[1, 0],
+                            0.0]).cast(dt).normalized()
+    else:
+        lambda1 = ti.Vector([tr, ti.sqrt(-gap)]) * 0.5
+        lambda2 = ti.Vector([tr, -ti.sqrt(-gap)]) * 0.5
+        A1r = A - lambda1[0] * ti.Matrix.identity(dt, 2)
+        A1i = -lambda1[1] * ti.Matrix.identity(dt, 2)
+        A2r = A - lambda2[0] * ti.Matrix.identity(dt, 2)
+        A2i = -lambda2[1] * ti.Matrix.identity(dt, 2)
+        v1 = ti.Vector([A2r[0, 0], A2i[0, 0], A2r[1, 0],
+                        A2i[1, 0]]).cast(dt).normalized()
+        v2 = ti.Vector([A1r[0, 0], A1i[0, 0], A1r[1, 0],
+                        A1i[1, 0]]).cast(dt).normalized()
+    eigenvalues = ti.Matrix.rows([lambda1, lambda2])
+    eigenvectors = ti.Matrix.cols([v1, v2])
```

Comment:
I think there is one thing that needs to be discussed. From my knowledge, taichi doesn't have such a datatype to represent complex numbers, which could be generated by eigensolvers. 

Currently, I use 2 numbers to store the real part and imaginary part of a complex number. That is to say, for a 2x2 matrix, function `ti.eig()` returns a 2x2 matrix v and a 4x2 matrix w. Each row of matrix v stores one eigenvalue represented by a 2d vector for its real and imaginary part. Each column of w stores one eigenvector with 2 entries each represented by two numbers for its real and imaginary part. I'm not sure whether this is the best way to handle this problem. Is there any advice on this? @k-ye

Reply:
I see. Yeah this sounds good, thanks! Could you write this as the docstring?